File: .virions_prepare.php
<?php

$path = __DIR__ . "/virions";

if (is_dir($path)) {
	remove($path);
}

if (!mkdir($path, 0777, true) && !is_dir($path)) {
	die("\n❌ Ошибка: Не удалось создать директорию '$path'. Проверьте права доступа.");
}

function downloadVirions(): void
{
	global $path;

	$virionsFile = __DIR__ . "/virions.txt";

	if (!file_exists($virionsFile) || !($virions = file($virionsFile, FILE_IGNORE_NEW_LINES | FILE_SKIP_EMPTY_LINES))) {
		die("\n❌ Ошибка: Файл 'virions.txt' пуст или отсутствует. Проверьте наличие файла с ресурсами.");
	}

	echo "\n📥 Начало скачивания вирионов...";

	foreach ($virions as $virion) {
		[$id, $name] = array_map("trim", explode(";", $virion));
		$savePath = "$path/$name.phar";
		$url = "https://poggit.pmmp.io/r/$id";

		echo "\n📥 Скачиваю вирион '$name' (ID: $id) с $url...";

		if (!($data = @file_get_contents($url))) {
			echo "\n❌ Ошибка: Не удалось скачать '$name'. Проверьте подключение и ID ресурса.";
			continue;
		}

		if (file_put_contents($savePath, $data) === false) {
			echo "\n❌ Ошибка: Не удалось сохранить '$name'. Проверьте права доступа.";
			continue;
		}

		echo "\n✅ Вирион '$name' успешно скачан!";
	}

	echo "\n✅ Все вирионы успешно скачаны!";
}

function remove(string $dir, array $skip = []): void
{
	if (!is_dir($dir)) return;

	foreach (array_diff(scandir($dir), [".", ".."], $skip) as $item) {
		$path = "$dir/$item";
		is_dir($path) ? remove($path) : unlink($path);
	}

	@rmdir($dir);
}

function recursiveCopy(string $source, string $destination): void
{
	if (!is_dir($source)) {
		echo "\n❌ Ошибка: Исходная директория '$source' не существует.";
		return;
	}

	@mkdir($destination, 0777, true);

	foreach (array_diff(scandir($source), [".", ".."]) as $item) {
		$srcPath = "$source/$item";
		$destPath = "$destination/$item";
		is_dir($srcPath) ? recursiveCopy($srcPath, $destPath) : copy($srcPath, $destPath);
	}
}

function extractVirions(): void
{
	global $path;

	echo "\n📦 Начало распаковки вирионов...";

	foreach (glob("$path/*.phar") as $pharFile) {
		$name = basename($pharFile, ".phar");
		$destination = "$path/$name";

		echo "\n⚡ Распаковываю вирион '$name'...";

		try {
			$phar = new Phar($pharFile);
			$phar->extractTo($destination);
		} catch (Exception $e) {
			echo "\n❌ Ошибка при распаковке '$name': " . $e->getMessage();
			remove($destination);

			if (file_exists($pharFile)) {
				unlink($pharFile);
			}

			continue;
		}

		unlink($pharFile);
		remove($destination, ["src"]);
		recursiveCopy("$destination/src", $path);
		remove($destination);
	}

	echo "\n✅ Все вирионы успешно распакованы!";
}

function ending(): void
{
	global $path;

	$autoload = __DIR__ . "/.virions_autoload.php";
	copy($autoload, $path . "/autoload.php");
}

downloadVirions();
extractVirions();
ending();

File: .virions_prepare.py
import os
import shutil
import requests
from subprocess import run
from pathlib import Path

path = Path.cwd() / "virions"

def remove(directory: str, skip: list = []):
	directory = Path(directory)

	if not directory.is_dir():
		return

	for item in directory.iterdir():
		if item.name in skip:
			continue
		if item.is_dir():
			remove(item)
		else:
			item.unlink()

	# Проверяем, пуста ли папка после удаления пропущенных файлов
	if not any(directory.iterdir()):
		directory.rmdir()

def recursive_copy(source: str, destination: str):
	source = Path(source)
	destination = Path(destination)
	if not source.is_dir():
		print(f"\n❌ Ошибка: Исходная директория '{source}' не существует.")
		return

	destination.mkdir(parents=True, exist_ok=True)

	for item in source.iterdir():
		dest_path = destination / item.name
		if item.is_dir():
			recursive_copy(item, dest_path)
		else:
			shutil.copy2(item, dest_path)

def download_virions():
	virions_file = Path(os.getcwd()) / "virions.txt"

	if not virions_file.exists():
		print("\n❌ Ошибка: Файл 'virions.txt' отсутствует.")
		return

	with virions_file.open("r", encoding="utf-8") as f:
		virions = [line.strip() for line in f if line.strip()]

	if not virions:
		print("\n❌ Ошибка: Файл 'virions.txt' пуст.")
		return

	print("\n📥 Начало скачивания вирионов...")

	for virion in virions:
		try:
			id_, name = map(str.strip, virion.split(";"))
			save_path = path / f"{name}.phar"
			url = f"https://poggit.pmmp.io/r/{id_}"

			print(f"\n📥 Скачиваю вирион '{name}' (ID: {id_}) с {url}...")
			response = requests.get(url)
			response.raise_for_status()

			with save_path.open("wb") as f:
				f.write(response.content)

			print(f"\n✅ Вирион '{name}' успешно скачан!")
		except Exception as e:
			print(f"\n❌ Ошибка при скачивании '{name}': {e}")
			continue

	print("\n✅ Все вирионы успешно скачаны!")

def extract_virions():
	path = Path(os.getcwd()) / "virions"
	print("\n📦 Начало распаковки вирионов...")

	for phar_file in path.glob("*.phar"):
		name = phar_file.stem
		destination = path / name

		print(f"\n⚡ Распаковываю вирион '{name}'...")
		try:
			command = ["php", "-r", f"$phar = new Phar('{phar_file}'); $phar->extractTo('{destination}');"]
			run(command)

			phar_file.unlink()
			remove(destination, ["src"])
			recursive_copy(destination / "src", path)
			remove(destination)

			print(f"\n✅ Вирион '{name}' успешно распакован!")
		except Exception as e:
			print(f"\n❌ Ошибка при распаковке '{name}': {e}")
			remove(destination)

			if phar_file.exists():
				phar_file.unlink()

	print("\n✅ Все вирионы успешно распакованы!")

def final():
	autoload = Path.cwd() / ".virions_autoload.php"
	shutil.copy2(autoload, path / "autoload.php")

if path.is_dir():
	remove(path)

path.mkdir(exist_ok=True)

if __name__ == "__main__":
	download_virions()
	extract_virions()
	final()